sips/sip-021/sip-021-nakamoto.md

```diff
@@ -0,0 +1,1051 @@
+
+Today, each miner receives a tenure block budget, which places hard limits on
+how much CPU, RAM, and I/O their block can consume when evaluated.  In this
+proposal, each miner begins its tenure with a fixed budget, but Stackers may opt
+to increase that budget through a vote.  This is done to enable the miner to
+continue to produce blocks if the next burnchain block is late.
+
+To achieve this, Stackers each keep track of the elapsed wall-clock time since
+the start of the tenure.  Once the expected tenure time has passed (e.g. 10
+minutes), they vote to grant the miner an additional tenure execution budget.
+This is achieved by having Stackers collectively produce a `TenureExtension`
+Stacks transaction, which the miner would include into one of their blocks.
+Once this transaction has been processed, the tenure block budget expands.
+
+## Support for WSTS Schnorr Signatures
+
+This proposal calls for adding support for WSTS Schnorr signatures as a
+transaction authorization mode to Stacks.  This enables Stackers to collectively
+sign a Stacks transaction.  The specification for the transaction authorization
+is as follows:
+
+* The spending condition's hash mode must be **0x04**.
+* An additional spending condition variant specific to single-key Schnorr
+  signatures, which can only be used if the hash mode is 0x04.  The spending
+condition variant is encoded as follows:
+  * A 65-byte **Schnorr signature**, consisting of:
+    * The 33-byte signed x-coordinate of the signature curve point
+    * The 32-byte scalar derived from the hash of the generator raised to the
+      _k_th power and the message
+  * A 33-byte **compressed secp256k1 public key**, consisting of its signed
+    x-coordinate
+
+To verify the transaction signature, the node both verifies that the public
+key's hash160 is equal to the spending condition's 20-byte hash, and verifies
+that the Schnorr signature over the `presign-sighash` value of the transaction
+was generated by corresponding private key.
+
+The address of a principal who signs a transaction with the WSTS Schnorr
+signature algorithm is identical to that of a single-sig ECDSA signer.  They
+have the same address versions -- i.e. on mainnet, the address starts with `SP`,
+and on testnet, it starts with `ST`.
+
+## New Block Validation Rules
+
+In this proposal, a block is valid if and only if the following are true:
+
+* The block is well-formed
+    * It has the correct version and mainnet/testnet flag
+    * **(NEW)** Its header contains the same Bitcoin block hash as the Bitcoin
+      block that contains its tenure's block-commit transaction
+    * The transaction Merkle tree root is consistent with the transactions
+    * The state root hash matches the MARF tip root hash once all transactions
+      are applied
+    * **(NEW)** the block header has a valid ECDSA signature from the miner
+    * **(NEW)** the block header has a valid WSTS Schnorr signature from the set
```

Comment:
Should the validation method be specified here?

Reply:
I actually think not - the WSTS Schnorr signature has a defined validity that already is encompassed by this line. What would you make clearer?